### PR TITLE
Updated Dash Masternode RPC Calls

### PR DESCRIPTION
**Pull request by mayoreee**
- Added the following RPC calls:
            - masternode list
            - masternode outputs
            - masternode payments
            - masternode status
            - masternode winners
            - masternode count
        
- Added integration tests for each of the RPC calls above

### CODE REVIEW

File: json/src/lib.rs [hunk 1]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
...
+    #[serde(rename = "blockhash")]
+    pub block_hash: String,
+    pub amount: u32,
+    pub masternodes: Vec<Masternode>,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Deserialize, Serialize)]
+pub struct DMNState {
+    pub service: String,
+    #[serde(rename = "registeredHeight")]
```
Comment (QuantumExplorer) [inline json/src/lib.rs:2037]: #[serde(rename_all = "camelCase")]

File: json/src/lib.rs [hunk 2]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
     pub enabled: u32,
 }
 
+#[derive(Clone, PartialEq, Eq, Debug, Deserialize, Serialize)]
+pub struct GetMasternodeListJSON {
```
Comment (QuantumExplorer) [inline json/src/lib.rs:1990]: This should be renamed to Masternode

File: json/src/lib.rs [hunk 3]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
...
+    #[serde(rename = "proTxHash")]
+    pub pro_tx_hash: String,
```
Comment (QuantumExplorer) [inline json/src/lib.rs:1992]: This should be decoded to a [u8;32]

File: json/src/lib.rs [hunk 4]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
...
+    pub address: String,
```
Comment (QuantumExplorer) [inline json/src/lib.rs:1993]: This should be decoded to some type of a Address struct.

File: json/src/lib.rs [hunk 5]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
...
+    #[serde(rename = "registeredHeight")]
+    pub registered_height: u32,
+    #[serde(rename = "lastPaidHeight")]
+    pub last_paid_height: u32,
+    #[serde(rename = "PoSePenalty")]
+    pub pose_penalty: u32,
+    #[serde(rename = "PoSeRevivedHeight")]
+    pub pose_revived_height: u32,
+    #[serde(rename = "PoSeBanHeight")]
+    pub pose_ban_height: i32,
```
Comment (QuantumExplorer) [inline json/src/lib.rs:2046]: u32

File: json/src/lib.rs [hunk 6]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
...
+    #[serde(rename = "revocationReason")]
+    pub revocation_reason: u32,
+    #[serde(rename = "ownerAddress")]
+    pub owner_address: String,
+    #[serde(rename = "votingAddress")]
+    pub voting_address: String,
+    #[serde(rename = "payoutAddress")]
+    pub payout_address: String,
+    #[serde(rename = "pubKeyOperator")]
```
Comment (QuantumExplorer) [inline json/src/lib.rs:2055]: [u8;96] (I think)

File: json/src/lib.rs [hunk 7]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
...
+    pub pubkey_operator: String,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Deserialize, Serialize)]
+pub struct GetMasternodeStatusResult {
```
Comment (QuantumExplorer) [inline json/src/lib.rs:2060]: Rename to MasternodeStatus

File: json/src/lib.rs [hunk 8]
```diff
@@ -1986,6 +1986,91 @@ pub struct GetMasternodeCountResult {
...
+    pub outpoint: String,
```
Comment (QuantumExplorer) [inline json/src/lib.rs:2061]: Create outpoint struct which is a hash + u16 (or maybe u32)

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: json/src/lib.rs (hunk 3 above)
Comment (mayoreee) [reply json/src/lib.rs:1992]: Not sure why this is needed, but the server returns a `String` data type here. As in the RPC docs:

Screenshot 2022-07-28 at 23 33 46

Screenshot 2022-07-28 at 23 34 01

File: json/src/lib.rs (hunk 6 above)
Comment (mayoreee) [reply json/src/lib.rs]: payout_address? @QuantumExplorer

File: json/src/lib.rs (hunk 4 above)
Comment (mayoreee) [reply json/src/lib.rs:1993]: @QuantumExplorer The `address` field here is `IP:port`. Not sure what you implied by Address struct for this field:

Screenshot 2022-07-29 at 14 14 44

Comment (mayoreee) [reply json/src/lib.rs]: Added an address struct with two fields: `ip` and `port`

File: json/src/lib.rs [hunk 9]
```diff
@@ -2005,3 +2119,47 @@ where
...
+        match str_sequence.as_str() {
+        "WAITING_FOR_PROTX" => MasternodeState::MASTERNODE_WAITING_FOR_PROTX,
+        "POSE_BANNED" => MasternodeState::MASTERNODE_POSE_BANNED,
+        "REMOVED" => MasternodeState::MASTERNODE_REMOVED,
+        "OPERATOR_KEY_CHANGED" => MasternodeState::MASTERNODE_OPERATOR_KEY_CHANGED,
+        "PROTX_IP_CHANGED" => MasternodeState::MASTERNODE_PROTX_IP_CHANGED,
+        "READY" => MasternodeState::MASTERNODE_READY,
+        "ERROR" => MasternodeState::MASTERNODE_ERROR,
+        "UNKNOWN" => MasternodeState::UNKNOWN,
+        _ => panic!(),
```
Comment (QuantumExplorer) [inline json/src/lib.rs:2160]: Panicking isn't a great idea here, as a new Masternode status could panic all of platform :) Let's create a new State, which would be MasternodeState::NONRECOGNISED .

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

**Review by QuantumExplorer — APPROVED** [review]

**Pull request merged**